File: ios/ai-app-bridge-ios/Sources/AiAppBridgeIOS/Capture/MobileCaptureStore.swift
import Foundation

public struct CaptureInput {
    public var stream: String
    public var targetKey: String
    public var runtimeEpoch: String
    public var captureId: Int64
    public var timestampMs: Int64
    public var record: [String: Any]
    public var actionId: String?
    public var source: String
    public var stateKey: String?

    public init(
        stream: String,
        targetKey: String,
        runtimeEpoch: String,
        captureId: Int64,
        timestampMs: Int64,
        record: [String: Any],
        actionId: String? = nil,
        source: String = "sdk",
        stateKey: String? = nil
    ) {
        self.stream = stream
        self.targetKey = targetKey
        self.runtimeEpoch = runtimeEpoch
        self.captureId = captureId
        self.timestampMs = timestampMs
        self.record = record
        self.actionId = actionId
        self.source = source
        self.stateKey = stateKey
    }
}

public struct AppendReceipt {
    public var status: String
    public var accepted: Bool
    public var committed: Bool
    public var dropped: Bool
    public var deduplicated: Bool
    public var mobileFactId: String?
    public var reason: String?
}

public struct CaptureQuery {
    public var view: String
    public var stream: String
    public var sinceId: Int64?
    public var sinceMs: Int64?
    public var limit: Int?
    public var platform: String
    public var afterActionId: String?
    public var cursor: String?
    public var runtimeEpoch: String?
    public var targetKey: String?
    public var mobileFactId: String?

    public init(
        view: String,
        stream: String,
        sinceId: Int64? = nil,
        sinceMs: Int64? = nil,
        limit: Int? = nil,
        platform: String = "ios",
        afterActionId: String? = nil,
        cursor: String? = nil,
        runtimeEpoch: String? = nil,
        targetKey: String? = nil,
        mobileFactId: String? = nil
    ) {
        self.view = view
        self.stream = stream
        self.sinceId = sinceId
        self.sinceMs = sinceMs
        self.limit = limit
        self.platform = platform
        self.afterActionId = afterActionId
        self.cursor = cursor
        self.runtimeEpoch = runtimeEpoch
        self.targetKey = targetKey
        self.mobileFactId = mobileFactId
    }
}

public struct CaptureCoverage {
    public var status: String
    public var gap: Bool
    public var committed: Bool
}

public struct CaptureFactRef {
    public var mobileFactId: String
    public var stream: String
    public var captureId: Int64
    public var targetKey: String? = nil
    public var runtimeEpoch: String? = nil
    public var capturedAtMs: Int64? = nil
}

public struct CapturePage {
    public var ok: Bool
    public var type: String
    public var items: [[String: Any]]
    public var count: Int
    public var coverage: CaptureCoverage
    public var gap: Bool
    public var hasMore: Bool
    public var refs: [CaptureFactRef]
    public var values: [String: Any]
    public var nextCursor: String? = nil
    public var watermarkCursor: String? = nil
    public var runtimeEpoch: String? = nil
    public var targetKey: String? = nil
    public var storeGeneration: Int64? = nil
    public var throughWatermark: Int64? = nil
    public var reason: String? = nil
    public var window: [String: Any]? = nil

    static func unavailable(_ stream: String, _ reason: String) -> CapturePage {
        CapturePage(ok: false, type: stream, items: [], count: 0,
                    coverage: CaptureCoverage(status: "unavailable", gap: true, committed: false),
                    gap: true, hasMore: false, refs: [], values: [:], reason: reason)
    }
}

public struct CaptureWatermark {
    public var streams: [String: Int64]
}

public struct ClearReceipt {
    public var ok: Bool
    public var generation: Int64
    public var reason: String? = nil
}

public struct StreamStatus {
    public var count: Int
    public var ownedBytes: Int64
    public var dropped: Int64
    public var gap: Bool
}

public struct CaptureStoreStatus {
    public var persistent: Bool
    public var generation: Int64
    public var ownedBytes: Int64
    public var budgetBytes: Int64
    public var dropped: Int64
    public var streams: [String: StreamStatus]
    public var reason: String? = nil
    public var pendingRecords: Int = 0
    public var pendingBytes: Int = 0
}

public struct ByteBudgets {
    public var logs: Int
    public var network: Int
    public var events: Int
    public var state: Int

    public init(logs: Int = 256 * 1024, network: Int = 384 * 1024, events: Int = 256 * 1024, state: Int = 128 * 1024) {
        self.logs = logs
        self.network = network
        self.events = events
        self.state = state
    }

    public func total() -> Int { logs + network + events + state }

    public func bytes(for stream: String) -> Int {
        switch stream {
        case "logs": return logs
        case "network": return network
        case "events": return events
        case "state": return state
        default: return 0
        }
    }
}

public struct CountCaps {
    public var logs: Int
    public var network: Int
    public var events: Int
    public var state: Int

    public init(logs: Int = 4096, network: Int = 2048, events: Int = 4096, state: Int = 512) {
        self.logs = logs
        self.network = network
        self.events = events
        self.state = state
    }

    public func cap(for stream: String) -> Int {
        switch stream {
        case "logs": return logs
        case "network": return network
        case "events": return events
        case "state": return state
        default: return 0
        }
    }
}

public final class MobileCaptureStore {
    private struct PendingAppend {
        let input: CaptureInput
        let durability: String
        let completion: (AppendReceipt) -> Void
    }

    // This queue only bridges store opening; queries never read it. A receipt
    // is delivered after the persistent backend assigns the original identity.
    private let maxPendingRecords = 256
    private let maxPendingBytes = 1024 * 1024
    private var opening = false
    private var pending: [PendingAppend] = []
    private var pendingBytes = 0
    private var backend: SegmentedCaptureBackend?
    private let budgets: ByteBudgets
    private let caps: CountCaps
    private var attachmentLoss = false
    private var attachmentVersion = 0
    private var unavailableReason = "capture_store_unavailable"
    private let lock = NSLock()

    public init(budgets: ByteBudgets = ByteBudgets(), caps: CountCaps = CountCaps()) {
        self.budgets = budgets
        self.caps = caps
    }

    func beginOpening() {
        lock.lock()
        defer { lock.unlock() }
        precondition(!opening && backend == nil && pending.isEmpty)
        opening = true
        unavailableReason = "capture_store_opening"
    }

    // Lifecycle calls this after each open. The token prevents a late writer
    // callback from attaching a store that has already been stopped/replaced.
    func attachPersistentStore(_ store: SegmentedFactStore, directory: URL,
                               targetKey: String, runtimeEpoch: String) {
        lock.lock()
        attachmentVersion += 1
        let version = attachmentVersion
        backend?.invalidate()
        backend = nil
        opening = true
        unavailableReason = "capture_store_opening"
        lock.unlock()
        store.status { [self] status in
            lock.lock()
            guard version == attachmentVersion else { lock.unlock(); return }
            if status.state == .open, status.operation.isSuccess {
                do {
                    backend = try SegmentedCaptureBackend(store: store, directory: directory,
                        targetKey: targetKey, runtimeEpoch: runtimeEpoch,
                        epochStartSequence: status.nextSequence > 0 ? status.nextSequence - 1 : 0,
                        budgets: budgets, caps: caps, initialLoss: attachmentLoss, existingRecords: status.recordCount)
                    attachmentLoss = false
                } catch { unavailableReason = "capture_metadata_unavailable" }
            } else {
                unavailableReason = "capture_store_unavailable"
            }
            let queued = pending
            pending = []
            pendingBytes = 0
            opening = false
            if backend == nil && !queued.isEmpty { attachmentLoss = true }
            // Submit all startup writes before publishing the backend to a
            // concurrent query and its durable watermark barrier.
            let receipts = queued.map { entry in
                backend?.append(entry.input, durability: entry.durability) ?? rejected(unavailableReason)
            }
            lock.unlock()
            for (entry, receipt) in zip(queued, receipts) { entry.completion(receipt) }
        }
    }

    func detachPersistentStore(reason: String = "capture_store_unavailable") {
        lock.lock()
        attachmentVersion += 1
        backend?.invalidate()
        backend = nil
        opening = false
        unavailableReason = reason
        let queued = pending
        pending = []
        pendingBytes = 0
        if !queued.isEmpty { attachmentLoss = true }
        lock.unlock()
        for entry in queued { entry.completion(rejected(reason)) }
    }

    // Completion may run on the store writer, like SegmentedFactStore's own
    // callbacks. It must not synchronously wait for a disk query on that writer.
    public func append(_ record: CaptureInput, durability: String = "async",
                       completion: @escaping (AppendReceipt) -> Void) {
        lock.lock()
        if let backend {
            let receipt = backend.append(record, durability: durability)
            lock.unlock()
            completion(receipt)
            return
        }
        var reason = unavailableReason
        if opening {
            do {
                let data = try JSONSerialization.data(withJSONObject: record.record, options: [.sortedKeys])
                let byteCount = data.count + record.stream.utf8.count + record.targetKey.utf8.count
                    + record.runtimeEpoch.utf8.count + record.source.utf8.count
                    + (record.actionId?.utf8.count ?? 0) + (record.stateKey?.utf8.count ?? 0)
                if pending.count >= maxPendingRecords || byteCount > maxPendingBytes - pendingBytes {
                    reason = "capture_startup_queue_full"
                } else {
                    var snapshot = record
                    snapshot.record = try JSONSerialization.jsonObject(with: data) as! [String: Any]
                    pending.append(PendingAppend(input: snapshot, durability: durability, completion: completion))
                    pendingBytes += byteCount
                    lock.unlock()
                    return
                }
            } catch { reason = "invalid_capture_payload" }
        }
        attachmentLoss = true
        lock.unlock()
        completion(rejected(reason))
    }

    private func rejected(_ reason: String) -> AppendReceipt {
        AppendReceipt(status: "dropped", accepted: false, committed: false, dropped: true,
                      deduplicated: false, mobileFactId: nil, reason: reason)
    }

    public func mark(_ streams: [String]) -> CaptureWatermark {
        lock.lock()
        defer { lock.unlock() }
        return backend?.mark(streams) ?? CaptureWatermark(streams: [:])
    }

    public func query(_ query: CaptureQuery) -> CapturePage {
        lock.lock()
        let selected = backend
        let reason = unavailableReason
        lock.unlock()
        return selected?.query(query) ?? .unavailable(query.stream, reason)
    }

    public func status() -> CaptureStoreStatus {
        lock.lock()
        defer { lock.unlock() }
        return backend?.status() ?? CaptureStoreStatus(persistent: false, generation: 0,
            ownedBytes: 0, budgetBytes: Int64(budgets.total()), dropped: attachmentLoss ? 1 : 0,
            streams: Dictionary(uniqueKeysWithValues: ["logs", "network", "events", "state"].map {
                ($0, StreamStatus(count: 0, ownedBytes: 0, dropped: attachmentLoss ? 1 : 0, gap: attachmentLoss))
            }), reason: unavailableReason, pendingRecords: pending.count, pendingBytes: pendingBytes)
    }

    public func clear(_ scope: String = "all") -> ClearReceipt {
        lock.lock()
        defer { lock.unlock() }
        return backend?.clear(scope) ?? ClearReceipt(ok: false, generation: 0, reason: "capture_store_unavailable")
    }
}

File: android/ai-app-bridge-android/src/main/kotlin/io/github/mobileaidev/aiappbridge/android/capture/MobileCaptureStore.kt
package io.github.mobileaidev.aiappbridge.android.capture

import org.json.JSONObject
import io.github.mobileaidev.aiappbridge.android.SegmentedFactStore
import java.io.File

data class CaptureInput(
    val stream: String,
    val targetKey: String,
    val runtimeEpoch: String,
    val captureId: Long,
    val timestampMs: Long,
    val record: JSONObject,
    val actionId: String? = null,
    val source: String = "sdk",
    val stateKey: String? = null,
)

data class AppendReceipt(
    val status: String,
    val accepted: Boolean,
    val committed: Boolean,
    val dropped: Boolean,
    val deduplicated: Boolean,
    val mobileFactId: String?,
    val reason: String? = null,
)

data class CaptureQuery(
    val view: String,
    val stream: String,
    val sinceId: Long? = null,
    val sinceMs: Long? = null,
    val limit: Int? = null,
    val platform: String = "android",
    val afterActionId: String? = null,
    val cursor: String? = null,
    val runtimeEpoch: String? = null,
    val targetKey: String? = null,
    val mobileFactId: String? = null,
)

data class CaptureCoverage(
    val status: String,
    val gap: Boolean,
    val committed: Boolean,
)

data class CaptureFactRef(
    val mobileFactId: String,
    val stream: String,
    val captureId: Long,
    val targetKey: String? = null,
    val runtimeEpoch: String? = null,
    val capturedAtMs: Long? = null,
)

data class CapturePage(
    val ok: Boolean,
    val type: String,
    val items: List<JSONObject>,
    val count: Int,
    val coverage: CaptureCoverage,
    val gap: Boolean,
    val hasMore: Boolean,
    val refs: List<CaptureFactRef>,
    val values: Map<String, Any?> = emptyMap(),
    val nextCursor: String? = null,
    val watermarkCursor: String? = null,
    val runtimeEpoch: String? = null,
    val targetKey: String? = null,
    val storeGeneration: Long? = null,
    val throughWatermark: Long? = null,
    val reason: String? = null,
    val window: JSONObject? = null,
)

data class CaptureWatermark(
    val streams: Map<String, Long>,
)

data class ClearReceipt(
    val ok: Boolean,
    val generation: Long,
)

data class CaptureStoreStatus(
    val persistent: Boolean,
    val generation: Long,
    val ownedBytes: Long,
    val budgetBytes: Long,
    val dropped: Long,
    val streams: Map<String, StreamStatus>,
)

data class StreamStatus(
    val count: Int,
    val ownedBytes: Long,
    val dropped: Long,
    val gap: Boolean,
)

data class ByteBudgets(
    val logs: Int = 256 * 1024,
    val network: Int = 384 * 1024,
    val events: Int = 256 * 1024,
    val state: Int = 128 * 1024,
) {
    fun total(): Int = logs + network + events + state

    fun forStream(stream: String): Int = when (stream) {
        "logs" -> logs
        "network" -> network
        "events" -> events
        "state" -> state
        else -> 0
    }
}

data class CountCaps(
    val logs: Int = 4096,
    val network: Int = 2048,
    val events: Int = 4096,
    val state: Int = 512,
) {
    fun forStream(stream: String): Int = when (stream) {
        "logs" -> logs
        "network" -> network
        "events" -> events
        "state" -> state
        else -> 0
    }
}

class MobileCaptureStore(
    private val budgets: ByteBudgets = ByteBudgets(),
    private val caps: CountCaps = CountCaps(),
) {
    @Volatile private var backend: CaptureBackend = BoundedMemoryCaptureBackend(budgets, caps)

    /** Called while the mobile store opens. No scans or payload hydration on the App thread. */
    internal fun usePersistentStore(store: SegmentedFactStore, directory: File, targetKey: String, runtimeEpoch: String,
        epochStartSequence: Long, existingRecords: Long = 0) {
        if (backend is SegmentedCaptureBackend) return
        synchronized(lock) {
            if (backend is SegmentedCaptureBackend) return
            val startup = backend as BoundedMemoryCaptureBackend
            val pending = startup.pendingRecords()
            require(pending.all { it.targetKey == targetKey && it.runtimeEpoch == runtimeEpoch }) {
                "capture_startup_identity_mismatch"
            }
            val persistent = SegmentedCaptureBackend(store, directory, targetKey, runtimeEpoch, epochStartSequence, budgets, caps,
                initialLossStreams = startup.status().streams.filterValues { it.gap }.keys, existingRecords = existingRecords)
            // This callback runs on the writer. Enqueue without awaiting that same writer; its
            // query barrier proves commitment, and rejected appends retain their stream loss fence.
            pending.forEach { persistent.append(it, "async") }
            backend = persistent
        }
    }

    internal fun detachPersistentStore() {
        synchronized(lock) { backend = BoundedMemoryCaptureBackend(budgets, caps) }
    }
    private val lock = Any()

    fun append(record: CaptureInput, durability: String = "async"): AppendReceipt =
        synchronized(lock) { backend.append(record, durability) }

    fun mark(streams: List<String>): CaptureWatermark =
        synchronized(lock) { backend.mark(streams) }

    fun query(query: CaptureQuery): CapturePage {
        val selected = backend
        // Disk reads must not hold the App callback lock. The backend validates its generation
        // again at the end, so concurrent clear yields an explicit unavailable page.
        return if (selected is SegmentedCaptureBackend) selected.query(query)
        else synchronized(lock) { backend.query(query) }
    }

    fun status(): CaptureStoreStatus =
        synchronized(lock) { backend.status() }

    fun clear(scope: String = "all"): ClearReceipt =
        synchronized(lock) { backend.clear(scope) }
}

internal interface CaptureBackend {
    fun append(record: CaptureInput, durability: String): AppendReceipt
    fun mark(streams: List<String>): CaptureWatermark
    fun query(query: CaptureQuery): CapturePage
    fun status(): CaptureStoreStatus
    fun clear(scope: String): ClearReceipt
}
